Expand audit entry diff count check to sample/source details update and to source bulk update dialog

cnathe opened the pull request:
#### Rationale
In the related PR from last week, I added a check in the sample bulk update test component save button click to verify the last audit entry event newRecordMap diff change count. This set of PRs moves that code to the AuditLogHelper class and extends it to be checked for both sample and source/registry bulk update and single sample/source overview detail update.

#### Related Pull Requests
- https://github.com/LabKey/limsModules/pull/1311
- https://github.com/LabKey/limsModules/pull/1332

#### Changes
- move audit entry diff count check helper code from SMTestUtils to AuditLogHelper
- EntityBulkUpdateDialog to check audit entry diff by default for clickUpdate()
- DetailTableEdit to check audit entry diff by default for clickSave

cnathe committed:
Add SMTestUtils.checkSampleTimelineAuditEventDiffCountForLastTransaction() for sample details edit save button click

cnathe committed:
Add option to skip changeCounter check for SampleDetailsPanel

cnathe committed:
Move check audit log diff event helpers from SMTestUtils to AuditLogHelper

cnathe committed:
Merge branch 'develop' into fb_sampleDetailEditAuditCheck

cnathe committed:
Move checkTimelineAuditEventDiffCountForLastTransaction from SampleOverviewPage to DetailTableEdit to check in sample and source/registry detail edit cases

cnathe committed:
move usages of getAuditEventNameFromURL() instead of hardcoding "SampleTimelineEvent"

cnathe committed:
Add audit diff checks for EntityBulkUpdateDialog.clickUpdate for both samples and sources/data classes

cnathe committed:
mist import cleanup

cnathe commented on src/org/labkey/test/util/AuditLogHelper.java:
```diff
@@ -69,6 +76,116 @@ public DataRegionTable goToAuditEventView(String eventType)
+            // filter out SampleStateLabel as that is not a change, it is added for display purposes
+            dataChanges = Stream.of(dataChanges).filter(s -> !s.toLowerCase().startsWith("samplestatelabel=")).toArray(String[]::new);
+            BaseWebDriverTest.getCurrentTest().log("Audit record data changes diff count check: " + dataChangesStr);
+            BaseWebDriverTest.getCurrentTest().checker().verifyEquals("Audit record data changes did not include the expected number of diffs, expected " + expectedDiffCount + " but was " + dataChanges.length + ": " + dataChangesStr,
```
@labkey-tchad You mentioned via chat that `BaseWebDriverTest.getCurrentTest()` isn't correct here because of the test scoping. Do you have a recommended approach here? Should I just throw or should I pass through the BaseWebDriverTest in some way?

labkey-tchad replied:
These methods shouldn't be `static`. Our standard pattern is that `Helper` classes are concrete classes that are constructed with all of the context they will need. `Util` classes are collections of static methods like these.
An instance of `AuditLogHelper` has everything it needs for these methods to be non-static, you'd just need to change `_wrapper` to be a `BaseWebDriverTest`; that would give easy access to the `checker()`.

cnathe replied:
ah, ok perfect. that makes sense. thank you

cnathe replied:
@labkey-tchad Trey, I'm going to need some help with this. I am able to change the _wrapper in AuditLogHelper to type BaseWebDriverTest, that part makes sense. However, I am still unsure how I can then use the AuditLogHelper inside of the component class such as EntityBulkUpdateDialog or DetailTableEdit. I don't see a way that they have access to the BaseWebDriverTest. It seems like what I'd really like is for that AuditLogHelper _wrapper to be of type WebDriverWrapper, but then I wouldn't have access to the checker().

labkey-tchad replied:
In that case, I'd say to just to a normal assertion. Making the `checker` available in a place like this might require some pretty significant refactoring and redesigning of our component patterns. Probably not worth the trouble at this time but something to consider in the future.

cnathe replied:
Gotcha, that makes sense to me. I'll make it an assertEquals(). Thank you.

labkey-tchad commented on src/org/labkey/test/util/AuditLogHelper.java:
```diff
@@ -69,6 +76,116 @@ public DataRegionTable goToAuditEventView(String eventType)
+    {
+        SelectRowsCommand cmd = new SelectRowsCommand("auditLog", auditEventName);
+        cmd.setColumns(columnNames);
+        cmd.addFilter("ProjectId/DisplayName", projectName, Filter.Operator.EQUAL);
```
This filter should use the project path, not the display name. We don't, generally, set custom container display names but we might some day.
```suggestion
        cmd.addFilter("ProjectId/Path", projectName, Filter.Operator.EQUAL);
```

cnathe replied:
I agree. I was just moving this method over from SMTestUtils, but I'll make this change.

labkey-tchad commented on src/org/labkey/test/util/AuditLogHelper.java:
```diff
@@ -69,6 +76,116 @@ public DataRegionTable goToAuditEventView(String eventType)
+     * @throws IOException Can be thrown by the SelectRowsCommand.
+     * @throws CommandException Can be thrown by the SelectRowsCommand.
+     */
+    public static SelectRowsResponse getAuditLogsFromLKS(String projectName, String folderName, String auditEventName, List<String> columnNames,
+                                                         List<Filter> filters, @Nullable Integer maxRows) throws IOException, CommandException
+    {
```
This should just take the containerPath as a single parameter. Having separate projectName and folderName is confusing when they are equal; technically, you could have a subfolder with the same name as the project.
```suggestion
    public static SelectRowsResponse getAuditLogsFromLKS(String containerPath, String auditEventName, List<String> columnNames,
                                                         List<Filter> filters, @Nullable Integer maxRows) throws IOException, CommandException
    {
        String projectName = StringUtils.strip(containerPath, " /").split("/")[0];
```

cnathe replied:
done

labkey-tchad commented on src/org/labkey/test/util/AuditLogHelper.java:
```diff
@@ -69,6 +76,116 @@ public DataRegionTable goToAuditEventView(String eventType)
+            String[] dataChanges = dataChangesStr != null ? dataChangesStr.split("&") : new String[0];
+            // filter out SampleStateLabel as that is not a change, it is added for display purposes
+            dataChanges = Stream.of(dataChanges).filter(s -> !s.toLowerCase().startsWith("samplestatelabel=")).toArray(String[]::new);
+            BaseWebDriverTest.getCurrentTest().log("Audit record data changes diff count check: " + dataChangesStr);
```
You can use `TestLogger.log` for logging from a static context.
```suggestion
            TestLogger.log("Audit record data changes diff count check: " + dataChangesStr);
```

cnathe replied:
done

cnathe committed:
Code review feedback:
- batch update dialog to use title rowCount by default for audit event row count check
- adjustChangeCounter() instead of a setChangeCounter()
- AuditLogHelper to not use static methods and to take containerPath instead of projectName plus folderName

labkey-tchad commented on src/org/labkey/test/util/AuditLogHelper.java:
```diff
@@ -69,6 +80,117 @@ public DataRegionTable goToAuditEventView(String eventType)
+        filters.forEach(cmd::addFilter);
+        if (maxRows != null)
+            cmd.setMaxRows(maxRows);
+        if (!containerPath.startsWith("/"))
+            containerPath = "/" + containerPath;
```
FYI, the Java client api cares whether the containerPath has a leading slash. It handles either case.

cnathe replied:
Thanks, removed.

labkey-tchad commented on src/org/labkey/test/util/AuditLogHelper.java:
```diff
@@ -69,6 +80,117 @@ public DataRegionTable goToAuditEventView(String eventType)
+        return null;
+    }
+
+    public static boolean isSamplesRoute()
+    {
+        URL url = BaseWebDriverTest.getCurrentTest().getURL();
+        if (url != null)
```
Doe these route checking methods need to be static? Want to avoid using `BaseWebDriverTest.getCurrentTest()`.
```suggestion
    public boolean isSamplesRoute()
    {
        URL url = _wrapper.getURL();
        if (url != null)
```

cnathe replied:
No need to be static anymore. Changed.

labkey-tchad commented on src/org/labkey/test/util/AuditLogHelper.java:
```diff
@@ -69,6 +80,117 @@ public DataRegionTable goToAuditEventView(String eventType)
+            cmd.setMaxRows(maxRows);
+        if (!containerPath.startsWith("/"))
+            containerPath = "/" + containerPath;
+        return cmd.execute(WebTestHelper.getRemoteApiConnection(), containerPath);
```
For consistency, use the connection supplier in this class rather than the `WebTestHelper` one.
```suggestion
        return cmd.execute(_connectionSupplier.get(), containerPath);
```

cnathe replied:
Done

cnathe committed:
Merge branch 'develop' into fb_sampleDetailEditAuditCheck

cnathe committed:
CR feedback
- remove static from methods in AuditLogHelper
- replace SMTestUtils code with AuditLogHelper now that those methods are moved

labkey-danield commented on src/org/labkey/test/util/AuditLogHelper.java:
```diff
@@ -69,6 +77,112 @@ public DataRegionTable goToAuditEventView(String eventType)
+    public void checkTimelineAuditEventDiffCount(String containerPath, String auditEventName, List<Integer> expectedDiffCounts) throws IOException, CommandException
+    {
+        Integer maxRows = expectedDiffCounts.size();
+        List<Map<String, Object>> events = getAuditLogsFromLKS(containerPath, auditEventName, List.of("NewRecordMap"), Collections.emptyList(), maxRows).getRows();
```
Should you assert the the events count matches the expectedDiffsCounts?

cnathe replied:
yup, added.

cnathe committed:
EntityBulkUpdateDialog to account for source newRecordMap including name by default

cnathe committed:
AuditLogHelper add assert for checking number of events

cnathe committed:
Merge branch 'develop' into fb_sampleDetailEditAuditCheck
